### PR TITLE
surface serialization error

### PR DESCRIPTION
**Pull request by lemony312**
solution to 
https://github.com/zio/zio-kafka/issues/1331

### CODE REVIEW

File: zio-kafka-test/src/test/scala/zio/kafka/serde/SerializerSpec.scala [hunk 1]
```diff
@@ -2,7 +2,7 @@ package zio.kafka.serde
 
 import org.apache.kafka.common.header.internals.RecordHeaders
 import zio.test.Assertion._
-import zio.test._
+import zio.test.{ test, _ }
```
Comment (erikvanoosten) [inline zio-kafka-test/src/test/scala/zio/kafka/serde/SerializerSpec.scala:5]: This shouldn't be necessary. Does it still compile after reverting this line?

File: zio-kafka/src/main/scala/zio/kafka/serde/Serializer.scala [hunk 2]
```diff
@@ -20,7 +20,7 @@ trait Serializer[-R, -T] {
    * Create a serializer for a type U based on the serializer for type T and a mapping function
    */
   def contramap[U](f: U => T): Serializer[R, U] =
-    Serializer((topic, headers, u) => serialize(topic, headers, f(u)))
+    Serializer((topic, headers, u) => ZIO.attempt(f(u)).flatMap(serialize(topic, headers, _)))
```
Comment (erikvanoosten) [inline zio-kafka/src/main/scala/zio/kafka/serde/Serializer.scala:23]: Could you make the same change for `contramapM` please?

**Review by erikvanoosten — COMMENTED** [review]
This won't be good for performance, regardless, it looks like a good idea. Would you address the small comments?

**Comment by svroonland** [comment]
> This won't be good for performance, regardless, it looks like a good idea. 

That's my concern as well, as this is executed for every message. Would it be better to add a warning to the docs that exceptions are not handled for this reason and result in a defect + refer to `contramapM` if you want error handling?

**Comment by erikvanoosten** [comment]
> > This won't be good for performance, regardless, it looks like a good idea.
> 
> That's my concern as well, as this is executed for every message. Would it be better to add a warning to the docs that exceptions are not handled for this reason and result in a defect + refer to `contramapM` if you want error handling?

Another option is to accept the performance hit and recommend people to do (de)serialization in their own code. We can even give some example code how to do this efficiently (especially taking chunks into account). WDYT?

**Comment by svroonland** [comment]
The Consumer already deserializes in chunks, not sure how that could be done more efficiently in user code, but probably I'm missing something. What did you have in mind?

**Comment by erikvanoosten** [comment]
Yes, that is the case. The point is more that if you run (de)serialization in your own code you should take chunking into account.
Still, it could go faster because you can make more assumptions. In particular, it is often okay to convert the entire chunk in a single ZIO. This saves a lot of zio-flatmapping.

**Comment by lemony312** [comment]
It's not ideal for performance but the code base already wraps serializers in ZIO.attempt for the java one
https://github.com/zio/zio-kafka/blob/master/zio-kafka/src/main/scala/zio/kafka/serde/Serializer.scala#L70

this is partially why the bug was hard to diagnose - it looks as though the error channel is strongly typed.

File: zio-kafka/src/main/scala/zio/kafka/serde/Serializer.scala (hunk 2 above)
Comment (lemony312) [reply zio-kafka/src/main/scala/zio/kafka/serde/Serializer.scala:23]: `contramapM` takes in a ZIO that can fail with throwable - do you think its worth wrapping?

Comment (erikvanoosten) [reply zio-kafka/src/main/scala/zio/kafka/serde/Serializer.scala:23]: Yes, that is good question. If I read https://zio.dev/guides/migrate/zio-2.x-migration-guide#guidelines-for-library-authors point 2 right, then yes we should do this.

**Comment by erikvanoosten** [comment]
> > This won't be good for performance, regardless, it looks like a good idea.
> 
> That's my concern as well, as this is executed for every message. Would it be better to add a warning to the docs that exceptions are not handled for this reason and result in a defect + refer to `contramapM` if you want error handling?

@svroonland If I read https://zio.dev/guides/migrate/zio-2.x-migration-guide#guidelines-for-library-authors point 2 right, then we should do the same wrapping in `contramapM`.

**Comment by ytalashko** [comment]
Just leaving my 2c, in case it is gonna be helpful, otherwise ignore it:
I like the suggestion in https://github.com/zio/zio-kafka/pull/1332#issuecomment-2363001954.
Basically, the error is happening in case the mapping function produces it.
The comments seem to be discussing two options to tackle that:
1) Lib handles exceptions for user, but this hurts performance.
2) It is users responsibility to not `throw` exceptions.

With first option, library works faster for users not throwing exceptions (which seems like a general case, at least my assumption) and produces defect in case of bugs (I say "bugs", 'cause if your code throws exceptions, you should ensure they are handled, and also in functional programming throwing exceptions is kinda an anti pattern). I believe `ZIO.map` works exactly in the same way. I also believe, it is a right way (produce defect in case of bug), 'cause otherwise performance hit is taken.

With second option, library works slower for users not throwing exceptions (which seems like a general case, at least my assumption) and handles user's bugs. This way library hurts users who not having bugs, which seems to me like a bad thing. Having API method taking pure function, you (at least I) assume it is not doing some additional job by trying to handle an exception. In case my mapping function throws errors, I would look for something that can take that into account (e.g.  `contramapM`) gives an ability to express that. Even, from the [issue description](https://github.com/zio/zio-kafka/issues/1331) it seems to be a bug to me, and having a defect in case of bugs is seems expected.

**Comment by svroonland** [comment]
Agreed with @ytalashko. As long as exceptions in `f` result in a died zio that is surfaced to user code through zio-kafka, this is acceptable to me and consistent with how ZIO's map and flatMap behave. If it's not surfaced we should see if we can fix that in the Producer. We can add a comment to `Serializer#contramap` warning about this and recommending to use `contramapM` if you expect errors in your `f`.

What do you think @erikvanoosten?

**Comment by erikvanoosten** [comment]
> Agreed with @ytalashko. As long as exceptions in `f` result in a died zio that is surfaced to user code through zio-kafka, this is acceptable to me and consistent with how ZIO's map and flatMap behave. If it's not surfaced we should see if we can fix that in the Producer. We can add a comment to `Serializer#contramap` warning about this and recommending to use `contramapM` if you expect errors in your `f`.
> 
> What do you think @erikvanoosten?

Yes, I agree. Thanks @ytalashko.

**Comment by erikvanoosten** [comment]
@lemony312 I am going to close this PR as per the discussion above.
For documenting the behavior we can create another PR.
Thanks for your effort. It didn't result in a code change, but at least we have some clarity.

**Pull request closed without merging**